### ISSUE TITLE
Please improve error message on deserialization problem

### ISSUE THREAD

**Issue by macias** [dotnet/aspnetcore#50439]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe the problem.

Consider such type:
```
public sealed class CategorySaveRequest
{
    public required List<Category> Categories { get; set; } 
}
```

I could force C# compiler to accept `null` as `Categories`. When doing pure serialization/deserialization it does not cause problem, but when I send a corrupted (with `null`) data to controller, with method like:

```
public  ValueTask<Result<CategorySaveResponse>> SaveCategoriesAsync([FromBody] CategorySaveRequest request)
```

I will get an error:

> {"type":"https://tools.ietf.org/html/rfc7231#section-6.5.1","title":"One or more validation errors occurred.","status":400,"traceId":"00-04c1e53cf1ace50dfdea3e139768c874-15202f87d156e377-00","errors":{"Categories":["The Categories field is required."]}}

The error message indicates that the code knows perfectly well which type to instantiate, however pointing out only member name causes the message to be more vague than someone would like to receive (just imagine 10 request types with `Categories` member).

### Describe the solution you'd like

Please enrich message error, something along the lines:

> {"type":"https://tools.ietf.org/html/rfc7231#section-6.5.1","title":"One or more validation errors
occurred.","status":400,"traceId":"00-04c1e53cf1ace50dfdea3e139768c874-15202f87d156e377-00","errors":{"Categories":["The Categories property of type CategorySaveRequest is required."]}}

*Side note: term "field" in original message is incorrect, but this issue is not so important.*

**Comment by macias** [comment]
@captainsafia The framework is NET 7.0. Sample code -- please take a look here https://codeberg.org/macias/Demo/src/branch/master/AspNetVagueErrorDeserialization

This is a bit modified template Blazor WebAsm project. Please run it, open web page for it, click on "Fetch data" menu (from left pane), open DevTools pane in the browser, and then click on button "POST". You will get error message saying:

> {"type":"https://tools.ietf.org/html/rfc7231#section-6.5.1","title":"One or more validation errors occurred.","status":400,"traceId":"00-f1f32066307a46f841dd1dbbd037fdff-997ff44e913b8e82-00","errors":{"Summary":["The Summary field is required."]}}

So, as my report states, the "parent" type name is missing here.

**Comment by captainsafia** [comment]
@macias You can modify the behavior of the generated response by configuring `InvalidModelStateResponseFactory` via the `ConfigureApiBehaviorOptions` extension method. You should be able to modify the keys and values for the error message from there.

You can also try returning a custom `ValidationProblemResult` from the action with your desired values but the former option might be more generic.

**Comment by macias** [comment]
@captainsafia thank you for the reply. This is useful of course but my point is to make it provide information out of the box. Or in other words, the defaults for any library (IMHO) should be useful, informative one, so there would be no need for tweaking right from the start. And this case falls into this category.

**Comment by captainsafia** [comment]
@macias Yep, I understand your original intent. The challenge here is that it's a matter of opinion whether the format desired above should be standard out-of-the-box. The fact that we provide an API for customization here gives us the flexibility to enable people to what's right for their use case without taking on the burden of implementing the one right approach inbox.

When possible, we always try to point people to the customization toggles that are available, instead of modifying the behavior of the framework, which can be breaking in some cases.

**Comment by macias** [comment]
Just to clarify, this is not about format, but coordinates. Maybe an analogy, you compile C# code, and you get this error:

```
Cannot assign int to variable x.
```

How helpful is it? Or would it be better:

```
File X, line 5, column 17: Cannot assign int to variable x.
```

? 

Same here. The coordinates are missing, so as in the first example I have to actually hunt down occurrences, instead of put a finger immediately at the problematic piece. This day and night experience.

**Comment by captainsafia** [comment]
Yep, my point being is that the difference between:

> {"Categories":["The Categories field is required."]}}

and 

> {"Categories":["The Categories property of type CategorySaveRequest is required."]}}

from a framework perspective is in the formatting of the error message which can be controlled by the `InvalidModelStateResponseFactory` API.
